### ISSUE TITLE
FIREBREAK IDEA - EKS Spike & Presentation

### ISSUE THREAD

**Issue by ollieanwyll** [ministryofjustice/cloud-platform#96]
## Overview

With EKS now officially launched, the team needs to become accustomed to its functionality, potential benefits and the integration and configuration process.

It looks like EKS is currently not available in London or Ireland, but it will be soon, and this spike can be completed in any region. 

https://aws.amazon.com/eks/

Targets:
1) Get as familiar with EKS as possible and figure out how it would be used within The Cloud Platform.

2) Make and perform a Presentation/Demo to share the Spike's knowledge with the team.

## Questions to be investigated

- [x] What methods of authentication are available? Can we use GitHub and OIDC?
- [x] Can EKS be added to our current Terraform resources, and if so with how much effort?
- [x] Are any backup or disaster recovery processes required with EKS?
- [x] Are we doing anything with Kops that is unachievable with EKS (except OIDC)?
- [x] Do our existing cluster components and services play nicely with EKS? external-dns, ingress-controller, Tiller
- [x] Do existing user-facing tools play nicely with EKS? kubectl, helm, kubectx, kubenx
- [x] Do user-facing cluster features work properly? kubectl exec, kubectl logs, kubectl proxy, etc
- [x] What is the monitoring like? Can you only use CloudWatch? Does CloudWatch monitor nodes/pods? Can you use third-party monitoring tools like Prometheus?
- [x] Can we configure any options on `apiserver`, `controller-manager` or `scheduler` components of the masters?
- [x] Is `etcd` included in EKS and if so, what kind of access do we have? Can we at least get a connection with `etcdctl`?

**Comment by ollieanwyll** [comment]
I'd love to do this! :)

**Comment by kerin** [comment]
Unfortunately the EKS Preview has ended, so we don’t have access anymore (and neither does anyone else outside AWS).

**Comment by nayoa** [comment]
What methods of authentication are available? Can we use Github and OIDC.

From the documentation, no. You can only use IAM with Heptio's authentication to authenticate with with the EKS cluster. https://docs.aws.amazon.com/eks/latest/userguide/managing-auth.html

A workaround could potentially be looking into Amazon SSO with Github. However, negative here is you need a Microsoft Active Directory. https://aws.amazon.com/single-sign-on/

**Comment by nayoa** [comment]
What methods of authentication are available? Can we use GitHub and OIDC?
At the minute, IAM authentication only using heptio-authenticator-aws 

Can EKS be added to our current Terraform resources, and if so with how much effort?
EKS modules are available on Terraform to create clusters, delete clusters, provision worker nodes etc.

 Are any backup or disaster recovery processes required with EKS?
No. There is no mention of backup or diaster recovery in EKS docs.

Are we doing anything with Kops that is unachievable with EKS (except OIDC)?
Nothing that we've found so far. 

Do our existing cluster components and services play nicely with EKS? external-dns, ingress-controller, Tiller
- Tiller works perfectly
- external-dns works according to AWS but I struggled to get it to work
- Ingress-controller should in theory work

Do existing user-facing tools play nicely with EKS? kubectl, helm, kubectx, kubenx?
Yes

 Do user-facing cluster features work properly? kubectl exec, kubectl logs, kubectl proxy, etc
Yes

What is the monitoring like? 
EKS suggest to use Kubernetes dashboard to monitor Kubernetes specific things like pods and namespaces.
You are able to monitor worker nodes using CloudWatch

 Can we configure any options on apiserver, controller-manager or scheduler components of the 
 masters?
No as users we don't have access or visibility of the master nodes. 

Is etcd included in EKS and if so, what kind of access do we have? Can we at least get a connection with etcdctl
Looks of it no. In conversation with Ben Paton (AWS TAM)
